lib/Solver/STPBuilder.cpp

```diff
@@ -465,26 +465,29 @@ ExprHandle STPBuilder::getInitialRead(const Array *root, unsigned index) {
   return vc_readExpr(vc, getInitialArray(root), bvConst32(32, index));
 }
 
-::VCExpr STPBuilder::getArrayForUpdate(const Array *root, 
+::VCExpr STPBuilder::getArrayForUpdate(const Array *root,
                                        const UpdateNode *un) {
+  ::VCExpr un_expr;
+  std::vector<const UpdateNode *> update_nodes;
+  for (; un && !_arr_hash.lookupUpdateNodeExpr(un, un_expr);
+       un = un->next.get()) {
+    update_nodes.emplace_back(un);
+  }
   if (!un) {
-      return getInitialArray(root);
+    un_expr = getInitialArray(root);
   }
-  else {
-      // FIXME: This really needs to be non-recursive.
-      ::VCExpr un_expr;
-      bool hashed = _arr_hash.lookupUpdateNodeExpr(un, un_expr);
-      
-      if (!hashed) {
-        un_expr =
-            vc_writeExpr(vc, getArrayForUpdate(root, un->next.get()),
-                         construct(un->index, 0), construct(un->value, 0));
 
-        _arr_hash.hashUpdateNodeExpr(un, un_expr);
-      }
-      
-      return un_expr;
+  for (std::size_t i = update_nodes.size(); i > 0;) {
+    --i;
+    un = update_nodes[i];
+
+    un_expr = vc_writeExpr(vc, getArrayForUpdate(root, un->next.get()),
```

Comment:
This should be now just referencing the last expr.
```suggestion
    un_expr = vc_writeExpr(vc, un_expr,
```
This way, we get rid of the recursive `getArrayForUpdate` call.

The reference counting can sometimes be a bit tricky but hopefully ASAN will tell ;)

Reply:
Yep, that was kind of the point of this PR... Good catch!